src/FsCheck/ArbMap.fs

```diff
@@ -11,31 +11,48 @@ type IArbMap =
     abstract ArbFor: Type -> Arbitrary<obj>
     abstract ArbFor<'T> : unit -> Arbitrary<'T>
 
+type private ArbMapInit = 
+| FromTypeClass of TypeClass.TypeClass<Arbitrary<obj>>
+| FromDiscover of (Type * ArbMap option)
```

Comment:
This could be two constructors or perhaps factory methods on `ArbMap`?

Reply:
I struggled with this constructor for a good while.

Static factories are out. There must be a reference to the ArbMap being initialized in the typeclass needed for initializing the ArbMap, else the typeclass wont resolve expected factory scenarios.

Here's an example of trying to split into two constructors ([comparison](https://github.com/farlee2121/FsCheck/compare/direct-arb-registration...farlee2121:FsCheck:separate-constructors?expand=1)). It compiles, but it throws 
> The initialization of an object or value resulted in an object or value being accessed recursively before it was fully initialized.

This recursion can't be fixed separating discover from registering the ArbMap instance as a config parameter([example](https://github.com/farlee2121/FsCheck/compare/direct-arb-registration...farlee2121:FsCheck:separate-constructors-2?expand=1)). If the ArbMap is not registered with the discover, then factories that require it will be filtered out by discover and lots of tests fail. 

We could solve this with a mutable field, but that didn't seem like a good option

Reply:
Short version: There is a tricky mutual dependence between the ArbMap instance and the typeclass it relies on which makes construction difficult